File: sdks/python/aegis_protocol_sdk/aegis_protocol/session.py
"""
aegis_protocol.session
--------------
Session state machine: INIT -> ACTIVE -> SUSPECT -> QUARANTINE -> FLUSHING -> RESYNC -> CLOSED
"""

import enum
import uuid
from typing import Callable, List, Optional

from aegis_protocol.policy import PolicyAction, PolicyDecision


class SessionState(enum.Enum):
    INIT       = "INIT"
    ACTIVE     = "ACTIVE"
    SUSPECT    = "SUSPECT"
    QUARANTINE = "QUARANTINE"
    FLUSHING   = "FLUSHING"
    RESYNC     = "RESYNC"
    CLOSED     = "CLOSED"


def state_str(state: SessionState) -> str:
    return state.value


FlushCallback = Callable[[str, str, List[str]], None]

# Warp score increments per decision action
WARP_DELTA: dict = {
    PolicyAction.ALLOW: -0.1,
    PolicyAction.FLAG:   0.5,
    PolicyAction.DENY:   1.0,
}
WARP_SUSPECT_THRESHOLD:    float = 1.0
WARP_QUARANTINE_THRESHOLD: float = 2.0


class Session:
    """
    Tracks the security state of an active agent session.
    Accumulates a "warp score" to detect anomalous behaviour.
    """

    def __init__(
        self,
        session_id: str,
        model_id: str,
        warp_threshold: float = 3.0,
        on_flush: Optional[FlushCallback] = None,
    ):
        self._session_id   = session_id
        self._model_id     = model_id
        self._threshold    = warp_threshold
        self._on_flush     = on_flush
        self._state        = SessionState.INIT
        self._warp         = 0.0
        self._tainted:     List[str] = []
        self._incident_id: str = ""

    # ------------------------------------------------------------------
    # Accessors
    # ------------------------------------------------------------------

    def state(self) -> SessionState:
        return self._state

    def warp_score(self) -> float:
        return self._warp

    def session_id(self) -> str:
        return self._session_id

    # ------------------------------------------------------------------
    # Lifecycle
    # ------------------------------------------------------------------

    def activate(self) -> None:
        self._require_state(SessionState.INIT)
        self._state = SessionState.ACTIVE

    def close(self) -> None:
        self._state = SessionState.CLOSED

    def reactivate(self) -> None:
        self._require_state(SessionState.RESYNC)
        self._state = SessionState.ACTIVE
        self._warp = 0.0
        self._tainted.clear()

    def complete_flush(self) -> None:
        self._require_state(SessionState.FLUSHING)
        if self._on_flush:
            self._on_flush(self._session_id, self._incident_id, list(self._tainted))
        self._state = SessionState.RESYNC

    # ------------------------------------------------------------------
    # Decision processing
    # ------------------------------------------------------------------

    def process_decision(self, decision: PolicyDecision, now: int) -> bool:
        """
        Apply a policy decision to the session.
        Returns True if the payload is allowed, False otherwise.
        """
        if self._state not in (SessionState.ACTIVE, SessionState.SUSPECT,
                               SessionState.QUARANTINE):
            return False

        allowed = decision.action == PolicyAction.ALLOW
        delta = WARP_DELTA.get(decision.action, 0.0)
        self._warp = max(0.0, self._warp + delta)

        if not allowed and decision.matched_rule_id:
            self._tainted.append(decision.matched_rule_id)

        self._advance_state()
        return allowed

    def _advance_state(self) -> None:
        if self._warp >= self._threshold:
            self._incident_id = str(uuid.uuid4())
            self._state = SessionState.FLUSHING
        elif self._warp >= WARP_QUARANTINE_THRESHOLD:
            self._state = SessionState.QUARANTINE
        elif self._warp >= WARP_SUSPECT_THRESHOLD:
            self._state = SessionState.SUSPECT
        else:
            if self._state in (SessionState.SUSPECT, SessionState.QUARANTINE):
                self._state = SessionState.ACTIVE

    def _require_state(self, expected: SessionState) -> None:
        if self._state != expected:
            raise RuntimeError(
                f"Expected state {expected.value} but current state is {self._state.value}"
            )

File: src/core/session.cpp
/*
 * Copyright 2026 Aegis Protocol Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 */

#include "uml001/core/session.h"
#include "uml001/crypto/crypto_utils.h"
#include "uml001/crypto/simple_hash_provider.h"
#include <chrono>
#include <algorithm>
#include <stdexcept>
#include <iostream>

namespace uml001 {

Session::Session(std::string session_id,
                 std::string peer_model_id,
                 float       warp_threshold,
                 FlushCallback on_flush)
    : session_id_(std::move(session_id))
    , peer_model_id_(std::move(peer_model_id))
    , warp_threshold_(warp_threshold)
    , on_flush_(std::move(on_flush))
    , state_(SessionState::INIT)
    , warp_score_(0.0f) {}

void Session::activate() {
    require_state(SessionState::INIT, "activate");
    transition(SessionState::ACTIVE, "handshake_complete");
}

bool Session::process_decision(const PolicyDecision& decision, uint64_t now_ms) {
    // Fail-closed logic for Terminal or Locked states
    if (state_ == SessionState::QUARANTINE ||
        state_ == SessionState::FLUSHING   ||
        state_ == SessionState::CLOSED) {
        return false;
    }

    // 1. Warp Score Calculation [E-8]
    float base_weight = 0.0f;
    switch (decision.action) {
        case PolicyAction::DENY:        base_weight = WARP_WEIGHT_DENY; break;
        case PolicyAction::FLAG:        base_weight = WARP_WEIGHT_FLAG; break;
        case PolicyAction::REQUIRE_MFA: base_weight = WARP_WEIGHT_MFA;  break;
        case PolicyAction::ALLOW:       base_weight = WARP_WEIGHT_ALLOW; break;
        default:                        base_weight = 0.0f; break;
    }

    // Apply risk multiplier from policy metadata
    float delta = base_weight * decision.risk_weight;
    warp_score_ = std::max(0.0f, warp_score_ + delta);

    // 2. Buffer payload for Entropy Management
    payload_buffer_.push_back(decision.payload_hash);
    if (payload_buffer_.size() > MAX_BUFFER) {
        payload_buffer_.pop_front();
    }

    // 3. State Transition Logic
    // Quarantine is triggered at 3x the threshold for absolute fail-stop
    if (warp_score_ >= (warp_threshold_ * 3.0f)) {
        transition(SessionState::QUARANTINE, "warp_critical_threshold");
        initiate_flush(now_ms);
        return false;
    } 
    // Suspect state triggered at 1x threshold
    else if (warp_score_ >= warp_threshold_ && state_ == SessionState::ACTIVE) {
        transition(SessionState::SUSPECT, "warp_elevated_risk");
    }
    // Recovery: Decay allows return to ACTIVE if behavior stabilizes
    else if (warp_score_ < (warp_threshold_ * 0.5f) && state_ == SessionState::SUSPECT) {
        transition(SessionState::ACTIVE, "warp_stabilized");
    }

    // Log the event for audit/transparency
    log_event({ now_ms, "POLICY_DECISION", decision.payload_hash,
                action_str(decision.action) + " weight=" + std::to_string(delta) });

    // 4. Entropy Flush Trigger
    // We flush on any DENY or when the entropy buffer is full to prevent probing
    if (decision.action == PolicyAction::DENY || payload_buffer_.size() >= MAX_BUFFER) {
        initiate_flush(now_ms);
    }

    return (state_ != SessionState::QUARANTINE && decision.action != PolicyAction::DENY);
}

void Session::initiate_flush(uint64_t now_ms) {
    // If we aren't already terminal, move to FLUSHING
    if (state_ != SessionState::QUARANTINE && state_ != SessionState::CLOSED) {
        transition(SessionState::FLUSHING, "entropy_flush_start");
    }

    // Generate unique incident ID using SHA256 context
    std::string salt = session_id_ + std::to_string(now_ms);
    std::string incident_id = crypto::SimpleHashProvider::instance().sha256(salt);
    
    std::vector<std::string> tainted(payload_buffer_.begin(), payload_buffer_.end());
    payload_buffer_.clear();
    
    if (on_flush_) {
        on_flush_(session_id_, incident_id, tainted);
    }

    // If we were just flushing (not quarantined), move to RESYNC to await re-handshake
    if (state_ == SessionState::FLUSHING) {
        transition(SessionState::RESYNC, "awaiting_resync");
    }
}

void Session::complete_flush() {
    require_state(SessionState::FLUSHING, "complete_flush");
    transition(SessionState::RESYNC, "flush_complete");
}

void Session::reactivate() {
    require_state(SessionState::RESYNC, "reactivate");
    warp_score_ = 0.0f; // Reset health on successful re-authentication
    payload_buffer_.clear();
    transition(SessionState::ACTIVE, "resync_successful");
}

void Session::transition(SessionState next, const std::string& reason) {
    if (state_ == next) return;

    std::string detail = state_str(state_) + " -> " +
                         state_str(next) + " reason=" + reason + 
                         " score=" + std::to_string(warp_score_);
    
    state_ = next;
    log_event({ 0, "STATE_TRANSITION", "", detail });
}

void Session::close() {
    transition(SessionState::CLOSED, "session_terminated");
}

bool Session::is_active() const {
    return (state_ == SessionState::ACTIVE || state_ == SessionState::SUSPECT);
}

void Session::require_state(SessionState expected, const std::string& op) const {
    if (state_ != expected) {
        throw std::runtime_error("Session error: " + op + " invalid in " + state_str(state_));
    }
}

void Session::log_event(SessionEvent e) {
    event_log_.push_back(std::move(e));
}

} // namespace uml001